cpp/src/arrow/compute/kernels/scalar_hash.cc

```diff
@@ -0,0 +1,221 @@
+                          var_length_buffer);
+  }
+
+  static Result<std::shared_ptr<ArrayData>> HashChild(const ArraySpan& array,
+                                                      const ArraySpan& child,
+                                                      LightContext* hash_ctx,
+                                                      MemoryPool* memory_pool) {
+    auto arrow_type = std::make_shared<ArrowType>();
+    auto buffer_size = child.length * sizeof(c_type);
+    ARROW_ASSIGN_OR_RAISE(auto buffer, AllocateBuffer(buffer_size, memory_pool));
+    ARROW_RETURN_NOT_OK(
+        HashArray(child, hash_ctx, memory_pool, buffer->mutable_data_as<c_type>()));
+    return ArrayData::Make(arrow_type, child.length,
+                           {array.GetBuffer(0), std::move(buffer)}, array.null_count);
+  }
+
+  static Status HashArray(const ArraySpan& array, LightContext* hash_ctx,
+                          MemoryPool* memory_pool, c_type* out) {
+    // KeyColumnArray objects are being passed to the hashing utility
+    std::vector<KeyColumnArray> columns(1);
+
+    auto type_id = array.type->id();
+    if (type_id == Type::EXTENSION) {
+      auto extension_type = checked_cast<const ExtensionType*>(array.type);
+      auto storage_array = array;
+      storage_array.type = extension_type->storage_type().get();
+      return HashArray(storage_array, hash_ctx, memory_pool, out);
+    }
+
+    if (type_id == Type::STRUCT) {
+      std::vector<std::shared_ptr<ArrayData>> child_hashes(array.child_data.size());
+      columns.resize(array.child_data.size());
+      for (size_t i = 0; i < array.child_data.size(); i++) {
+        auto child = array.child_data[i];
+        if (is_nested(child.type->id())) {
+          ARROW_ASSIGN_OR_RAISE(child_hashes[i],
+                                HashChild(array, child, hash_ctx, memory_pool));
+          ARROW_ASSIGN_OR_RAISE(columns[i], ToColumnArray(*child_hashes[i], hash_ctx));
+        } else {
+          ARROW_ASSIGN_OR_RAISE(columns[i], ToColumnArray(child, hash_ctx));
+        }
+      }
+      Hasher::HashMultiColumn(columns, hash_ctx, out);
+    } else if (is_list_like(type_id)) {
+      auto values = array.child_data[0];
+      ARROW_ASSIGN_OR_RAISE(auto value_hashes,
+                            HashChild(array, values, hash_ctx, memory_pool));
+      ARROW_ASSIGN_OR_RAISE(
+          columns[0], ToColumnArray(array, hash_ctx, value_hashes->buffers[1]->data()));
+      Hasher::HashMultiColumn(columns, hash_ctx, out);
+    } else {
+      ARROW_ASSIGN_OR_RAISE(columns[0], ToColumnArray(array, hash_ctx));
+      Hasher::HashMultiColumn(columns, hash_ctx, out);
+    }
+    return Status::OK();
+  }
+
+  static Status Exec(KernelContext* ctx, const ExecSpan& input_arg, ExecResult* out) {
+    if (input_arg.num_values() != 1 || !input_arg[0].is_array()) {
+      return Status::Invalid("FastHash currently supports a single array input");
+    }
```

Comment:
```suggestion
    DCHECK_EQ(input_arg.num_values(), 1);
    DCHECK(input_arg[0].is_array());
```

Reply:
These could be `DCHECK`s as we expect the function resolution already handled the argument number/type/shape.

Reply:
Fixed.